File: src/rtve_dl/ffmpeg.py
from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

from rtve_dl.log import debug, is_debug


def require_ffmpeg() -> None:
    if shutil.which("ffmpeg") is None:
        raise RuntimeError("ffmpeg not found on PATH")


def run_ffmpeg(args: list[str]) -> None:
    require_ffmpeg()
    base = ["ffmpeg", "-hide_banner", "-nostdin"]
    if is_debug():
        # Show progress for long downloads/mux operations.
        base += ["-loglevel", "warning", "-stats"]
        debug("ffmpeg " + " ".join(args))
    else:
        base += ["-loglevel", "error"]
    p = subprocess.run([*base, *args], text=True)
    if p.returncode != 0:
        raise RuntimeError(f"ffmpeg failed: {' '.join(args)}")


def is_valid_mp4(path: Path) -> bool:
    """
    Best-effort MP4 integrity check for cache-hit decisions.
    """
    if not path.exists() or path.stat().st_size == 0:
        return False

    ffprobe = shutil.which("ffprobe")
    if ffprobe:
        p = subprocess.run(
            [
                ffprobe,
                "-v",
                "error",
                "-select_streams",
                "v:0",
                "-show_entries",
                "format=duration",
                "-of",
                "default=nokey=1:noprint_wrappers=1",
                str(path),
            ],
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        if p.returncode != 0:
            return False
        out = (p.stdout or "").strip()
        return bool(out)

    # Fallback if ffprobe is unavailable.
    p = subprocess.run(
        ["ffmpeg", "-v", "error", "-i", str(path), "-f", "null", "-"],
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return p.returncode == 0


def download_to_mp4(input_url: str, out_mp4: Path, *, headers: dict[str, str] | None = None) -> None:
    out_mp4.parent.mkdir(parents=True, exist_ok=True)
    debug(f"download_to_mp4: {input_url} -> {out_mp4}")
    if out_mp4.exists():
        debug(f"cache hit mp4: {out_mp4}")
        return

    # Always download to a temporary file first, then atomically rename.
    part_mp4 = out_mp4.with_name(out_mp4.name + ".partial.mp4")

    # For direct MP4 URLs prefer curl resume; ffmpeg remux from a byte range is not
    # a safe "append" strategy for already-partial MP4 output files.
    if ".mp4" in input_url and shutil.which("curl") is not None:
        cmd: list[str] = [
            "curl",
            "--location",
            "--fail",
            "--silent",
            "--show-error",
            "--continue-at",
            "-",
            "--output",
            str(part_mp4),
            "--user-agent",
            (
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/144.0.0.0 Safari/537.36"
            ),
        ]
        if headers:
            for k, v in headers.items():
                cmd += ["--header", f"{k}: {v}"]
        cmd += [input_url]
        debug(" ".join(cmd))
        p = subprocess.run(cmd, text=True)
        if p.returncode == 0:
            part_mp4.replace(out_mp4)
            return
        debug(f"curl resume failed (exit {p.returncode}); falling back to ffmpeg: {input_url}")

    args: list[str] = ["-y"]
    if headers:
        # ffmpeg expects CRLF separated headers.
        hdr = "".join([f"{k}: {v}\r\n" for k, v in headers.items()])
        args += ["-headers", hdr]
    # For HLS this will remux; for MP4 it will copy. If it fails, user can pick another URL.
    args += ["-i", input_url, "-c", "copy", str(part_mp4)]
    run_ffmpeg(args)
    part_mp4.replace(out_mp4)


def mux_mkv(
    *,
    video_path: Path,
    out_mkv: Path,
    subs: list[tuple[Path, str, str]],
    subtitle_delay_ms: int = 0,
) -> None:
    """
    subs: list of (path, language, title). Codec will be SRT-in-MKV.
    """
    out_mkv.parent.mkdir(parents=True, exist_ok=True)
    debug(
        "mux_mkv: "
        f"video={video_path} out={out_mkv} subs={len(subs)} subtitle_delay_ms={subtitle_delay_ms}"
    )
    args: list[str] = ["-y", "-i", str(video_path)]
    subtitle_offset_sec = f"{subtitle_delay_ms / 1000.0:.3f}"
    for p, _lang, _title in subs:
        # Apply subtitle delay at mux stage only; keep cached SRT files unchanged.
        args += ["-itsoffset", subtitle_offset_sec, "-i", str(p)]

    # Map all streams: video+audio from input 0; then each subtitle input.
    args += ["-map", "0"]
    for i in range(1, 1 + len(subs)):
        args += ["-map", str(i)]

    # Copy primary A/V streams, re-encode subtitle inputs as SRT-in-MKV.
    args += ["-c:v", "copy", "-c:a", "copy", "-c:s", "srt"]

    # Attach metadata per subtitle stream.
    for idx, (_p, lang, title) in enumerate(subs):
        args += [f"-metadata:s:s:{idx}", f"language={lang}"]
        args += [f"-metadata:s:s:{idx}", f"title={title}"]

    args += [str(out_mkv)]
    run_ffmpeg(args)

File: rust/src/ffmpeg.rs
use crate::logging::debug;
use anyhow::{bail, Context, Result};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

fn require_tool(name: &str) -> Result<()> {
    let status = Command::new("sh")
        .arg("-lc")
        .arg(format!("command -v {name} >/dev/null"))
        .status()?;
    if !status.success() {
        bail!("{name} not found on PATH");
    }
    Ok(())
}

pub fn require_ffmpeg() -> Result<()> {
    require_tool("ffmpeg")
}

pub fn require_ffprobe() -> Result<()> {
    require_tool("ffprobe")
}

pub fn run_ffmpeg(args: &[String]) -> Result<()> {
    require_ffmpeg()?;
    debug(format!("ffmpeg {}", args.join(" ")));
    let status = Command::new("ffmpeg")
        .arg("-hide_banner")
        .arg("-nostdin")
        .arg("-loglevel")
        .arg(if crate::logging::is_debug() {
            "warning"
        } else {
            "error"
        })
        .args(args)
        .status()
        .context("run ffmpeg")?;
    if !status.success() {
        bail!("ffmpeg failed");
    }
    Ok(())
}

pub fn ffmpeg_encoders_text() -> Result<String> {
    require_ffmpeg()?;
    let output = Command::new("ffmpeg")
        .arg("-hide_banner")
        .arg("-encoders")
        .output()
        .context("ffmpeg -encoders")?;
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn pick_available_hevc_gpu_encoder() -> Result<Option<String>> {
    let encoders = ffmpeg_encoders_text()?;
    for name in [
        "hevc_videotoolbox",
        "hevc_nvenc",
        "hevc_qsv",
        "hevc_amf",
        "hevc_vaapi",
    ] {
        if encoders.contains(name) {
            return Ok(Some(name.to_string()));
        }
    }
    Ok(None)
}

pub fn probe_duration_seconds(path: &Path) -> Result<Option<f64>> {
    if !path.exists() || path.metadata()?.len() == 0 {
        return Ok(None);
    }
    require_ffprobe()?;
    let output = Command::new("ffprobe")
        .arg("-v")
        .arg("error")
        .arg("-show_entries")
        .arg("format=duration")
        .arg("-of")
        .arg("default=nokey=1:noprint_wrappers=1")
        .arg(path)
        .output()
        .context("ffprobe duration")?;
    if !output.status.success() {
        return Ok(None);
    }
    let out = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if out.is_empty() {
        return Ok(None);
    }
    Ok(out.parse::<f64>().ok().filter(|x| *x > 0.0))
}

pub fn is_valid_mp4(path: &Path) -> Result<bool> {
    Ok(probe_duration_seconds(path)?.is_some())
}

pub fn download_to_mp4(input_url: &str, out_mp4: &Path) -> Result<()> {
    out_mp4.parent().unwrap().mkdir_if_missing()?;
    if out_mp4.exists() && is_valid_mp4(out_mp4)? {
        return Ok(());
    }
    let part_mp4 = PathBuf::from(format!("{}.partial.mp4", out_mp4.display()));
    if input_url.contains(".mp4") {
        let status = Command::new("curl")
            .arg("--location")
            .arg("--fail")
            .arg("--silent")
            .arg("--show-error")
            .arg("--continue-at")
            .arg("-")
            .arg("--output")
            .arg(&part_mp4)
            .arg("--user-agent")
            .arg("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36")
            .arg(input_url)
            .status();
        if let Ok(status) = status {
            if status.success() && part_mp4.exists() && part_mp4.metadata()?.len() > 0 {
                fs::rename(&part_mp4, out_mp4)?;
                return Ok(());
            }
        }
    }
    let args = vec![
        "-y".to_string(),
        "-i".to_string(),
        input_url.to_string(),
        "-c".to_string(),
        "copy".to_string(),
        part_mp4.display().to_string(),
    ];
    run_ffmpeg(&args)?;
    fs::rename(&part_mp4, out_mp4)?;
    Ok(())
}

pub fn mux_mkv(
    video_path: &Path,
    out_mkv: &Path,
    subs: &[(PathBuf, String, String)],
    default_subtitle_title: Option<&str>,
    video_codec_mode: &str,
    hevc_device: &str,
    hevc_crf: i32,
    hevc_preset: &str,
) -> Result<()> {
    out_mkv.parent().unwrap().mkdir_if_missing()?;
    let mut args = vec![
        "-y".to_string(),
        "-i".to_string(),
        video_path.display().to_string(),
    ];
    for (path, _, _) in subs {
        args.push("-itsoffset".to_string());
        args.push("0.000".to_string());
        args.push("-i".to_string());
        args.push(path.display().to_string());
    }
    args.push("-map".to_string());
    args.push("0".to_string());
    for i in 1..=subs.len() {
        args.push("-map".to_string());
        args.push(i.to_string());
    }
    let default_idx = default_subtitle_title
        .and_then(|wanted| subs.iter().position(|(_, _, title)| title == wanted));
    for (idx, (_, lang, title)) in subs.iter().enumerate() {
        args.push(format!("-metadata:s:s:{idx}"));
        args.push(format!("language={lang}"));
        args.push(format!("-metadata:s:s:{idx}"));
        args.push(format!("title={title}"));
        if Some(idx) != default_idx {
            args.push(format!("-disposition:s:{idx}"));
            args.push("0".to_string());
        }
    }
    if let Some(idx) = default_idx {
        args.push(format!("-disposition:s:{idx}"));
        args.push("default".to_string());
    }

    match video_codec_mode {
        "copy" => {
            args.extend([
                "-c:v".to_string(),
                "copy".to_string(),
                "-c:a".to_string(),
                "copy".to_string(),
                "-c:s".to_string(),
                "srt".to_string(),
                out_mkv.display().to_string(),
            ]);
            run_ffmpeg(&args)
        }
        "hevc" => {
            if hevc_device == "cpu" {
                let mut cpu_args = args.clone();
                cpu_args.extend([
                    "-c:v".to_string(),
                    "libx265".to_string(),
                    "-crf".to_string(),
                    hevc_crf.to_string(),
                    "-preset".to_string(),
                    hevc_preset.to_string(),
                    "-c:a".to_string(),
                    "copy".to_string(),
                    "-c:s".to_string(),
                    "srt".to_string(),
                    out_mkv.display().to_string(),
                ]);
                return run_ffmpeg(&cpu_args);
            }
            if let Some(gpu) = pick_available_hevc_gpu_encoder()? {
                let mut gpu_args = args.clone();
                gpu_args.extend([
                    "-c:v".to_string(),
                    gpu,
                    "-c:a".to_string(),
                    "copy".to_string(),
                    "-c:s".to_string(),
                    "srt".to_string(),
                    out_mkv.display().to_string(),
                ]);
                if run_ffmpeg(&gpu_args).is_ok() {
                    return Ok(());
                }
            }
            let mut cpu_args = args;
            cpu_args.extend([
                "-c:v".to_string(),
                "libx265".to_string(),
                "-crf".to_string(),
                hevc_crf.to_string(),
                "-preset".to_string(),
                hevc_preset.to_string(),
                "-c:a".to_string(),
                "copy".to_string(),
                "-c:s".to_string(),
                "srt".to_string(),
                out_mkv.display().to_string(),
            ]);
            run_ffmpeg(&cpu_args)
        }
        other => bail!("unsupported video codec mode: {other}"),
    }
}

trait DirCreate {
    fn mkdir_if_missing(&self) -> Result<()>;
}

impl DirCreate for Path {
    fn mkdir_if_missing(&self) -> Result<()> {
        fs::create_dir_all(self).context("create directory")
    }
}
